statocles-site/blog/2024/12/25/jwst/index.markdown

```diff
@@ -0,0 +1,139 @@
+
+    # Make a gaussian extraction profile
+
+    $x = xvals($dims[1]);
+    $fwhm = 2.1; # pix
+    $gaussian = 0.04*exp(-0.5*(($x-15)/($fwhm/2.35))**2);
+    line $x, $gaussian; # Plot
+
+This code generates a 31 element gaussian function along the slit axis. Here is what `$gaussian` looks like:
+
+![image](https://github.com/user-attachments/assets/7c3bf9aa-b028-4511-83f2-e460fd0afcd8)
+
+(It does not look smooth as it is evaluated on a coarse pixel grid and is only a few pixels wide). If we can extract along the slit weighting by this gaussian we can get two things for the price of one: (1) eliminate the signal from the dark artefacts and (2) _optimally_ weight the spectrum according to its light distribution along the slit (this arises from the shape of the galaxy) to get the most signal:noise.
+
+Now we can do some _broadcasting_ and some _reduction_:
+
+    $profile_weighted = $jwst * $gaussian->clump(0); # Multiply through
+    $spec1d = $profile_weighted->mv(1,0)->sumover; # Extract by summing along columns
+    $spec1d = $spec1d/max($spec1d); # Normalise
+
+Recall `$jwst` has dimensions 435,31. We see another dimension trick, `$gaussian->clump(0)` adds a unit dimension at position 0 resulting in a dimension 1,31 ndarray. When we multiply these ndarrays together the second 31 element dimensions match, and the first  dimensions (435 and 1) are also matched. What happens is PDL implicitly expands the unit dimension by repeating it 435 times, and in the multiplication `$gaussian->clump(0)` behaves as a 435,31 ndarray. It is like multiplying two 435,31 ndarrays together. This ultimately results in the gaussian being multiplied through each column, which is then summed by the following line.
+
+This trick is known as 'broadcasting' and is one of the most powerful PDL features. You can see we have written some highly complex code to apply mathematical operations to the image as a function of x,y without writing a single loop! Broadcasting is powerful because it can be applied along any axis with suitable dimension tricks using functions such as `mv()` and `clump()`.  It is also extremely fast as it operates at speeds close to what would happen if the loops were written in C or FORTRAN. When using it I tend to need to experiment a bit to achieve what I want and look at the resulting dimensions. The main rule to remember is dimensions need to _match_, this happens when they are either the same size, or when one of them is of size unity. In the latter case the 'broadcasting' happens and you can think of it as growing the axis by repetition. This all happens in the PDL internals during the operation it is being applied to, and the ndarray does not actually get any bigger nor use more memory! The same goes for dimension manipulation with functions like `clump()` and `mv()`, the ndarray is not copied it is simply 'viewed differently'.
```

Comment:
Actually `clump` *does* copy, but it flows.

Reply:
should we _add_ `t` to line 99 to the list of mv() and clump() because it is correct for mv too even though we are replacing it

Reply:
I'd like to replace all `mv` with `t` and then just comment that for switching other dimensions, `mv` and `xchg` are available.

Reply:
The text still shows `clump` as "not copying", which is false.